File: nodecore-spv/src/main/java/veriblock/service/PendingTransactionContainer.java
package veriblock.service;

import nodecore.api.grpc.VeriBlockMessages;
import org.veriblock.sdk.models.Sha256Hash;
import spark.utils.CollectionUtils;
import veriblock.model.Transaction;
import veriblock.service.PendingTransactionContainer;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

public class PendingTransactionContainer {
    private final Map<String, ArrayList<Transaction>> pendingAddressTransaction = new ConcurrentHashMap<>();
    private final Map<Sha256Hash, Transaction> pendingTxIdTransaction = new ConcurrentHashMap<>();
    private final Map<Sha256Hash, VeriBlockMessages.TransactionInfo> confirmedTxIdTransactionReply = new ConcurrentHashMap<>();
    private final Set<Sha256Hash> transactionsForMonitoring = ConcurrentHashMap.newKeySet();

    public Set<Sha256Hash> getPendingTransactionsId() {
        Set<Sha256Hash> allPendingTx = new HashSet<>();
        allPendingTx.addAll(pendingTxIdTransaction.keySet());
        allPendingTx.addAll(transactionsForMonitoring);

        return allPendingTx;
    }

    public VeriBlockMessages.TransactionInfo getTransactionInfo(Sha256Hash txId) {
        if (confirmedTxIdTransactionReply.containsKey(txId)) {
            return confirmedTxIdTransactionReply.get(txId);
        }

        if (!pendingTxIdTransaction.containsKey(txId)) {
            transactionsForMonitoring.add(txId);
        }

        return VeriBlockMessages.TransactionInfo.newBuilder()
            .setConfirmations(0)
            .build();
    }

    public void updateTransactionInfo(VeriBlockMessages.TransactionInfo transactionInfo) {
        Sha256Hash txId = Sha256Hash.wrap(transactionInfo.getTransaction().getTxId().toByteArray());

        if (pendingTxIdTransaction.containsKey(txId)) {
            confirmedTxIdTransactionReply.put(txId, transactionInfo);

            if (transactionInfo.getConfirmations() > 0) {
                pendingTxIdTransaction.remove(txId);
                transactionsForMonitoring.remove(txId);
            }
        }
    }

    public boolean addTransaction(Transaction transaction) {
        ArrayList<Transaction> transactions = pendingAddressTransaction.get(transaction.getInputAddress().toString());

        if (CollectionUtils.isNotEmpty(transactions)) {
            transactions.add(transaction);
            pendingTxIdTransaction.put(transaction.getTxId(), transaction);
        } else {
            ArrayList<Transaction> newList = new ArrayList<>();
            newList.add(transaction);

            pendingTxIdTransaction.put(transaction.getTxId(), transaction);
            pendingAddressTransaction.put(transaction.getInputAddress().toString(), newList);
        }
        return true;
    }

    public Transaction getTransaction(Sha256Hash txId) {
        return pendingTxIdTransaction.get(txId);
    }

    public Long getPendingSignatureIndexForAddress(String address) {
        ArrayList<Transaction> transactions = pendingAddressTransaction.get(address);

        if(CollectionUtils.isNotEmpty(transactions)){
            return transactions.get(transactions.size() - 1).getSignatureIndex();
        }
        return null;
    }
}

File: nodecore-spv/src/main/java/veriblock/service/PendingTransactionContainer.kt
package veriblock.service

import org.veriblock.core.crypto.Sha256Hash
import veriblock.model.Transaction
import java.util.ArrayList
import java.util.HashSet
import java.util.concurrent.ConcurrentHashMap

class PendingTransactionContainer {
    private val pendingAddressTransaction: MutableMap<String, ArrayList<Transaction>> = ConcurrentHashMap()
    private val pendingTxIdTransaction: MutableMap<Sha256Hash, Transaction> = ConcurrentHashMap()
    private val confirmedTxIdTransactionReply: MutableMap<Sha256Hash, TransactionInfo> = ConcurrentHashMap()
    private val transactionsForMonitoring: MutableSet<Sha256Hash> = ConcurrentHashMap.newKeySet()

    fun getPendingTransactionIds(): Set<Sha256Hash> {
        val allPendingTx: MutableSet<Sha256Hash> = HashSet()
        allPendingTx.addAll(pendingTxIdTransaction.keys)
        allPendingTx.addAll(transactionsForMonitoring)
        return allPendingTx
    }

    fun getTransactionInfo(txId: Sha256Hash): TransactionInfo? {
        confirmedTxIdTransactionReply[txId]?.let {
            return it
        }
        if (!pendingTxIdTransaction.containsKey(txId)) {
            transactionsForMonitoring.add(txId)
        }
        return null
    }

    fun updateTransactionInfo(transactionInfo: TransactionInfo) {
        val transaction = transactionInfo.transaction
        if (pendingTxIdTransaction.containsKey(transaction.txId) || transactionsForMonitoring.contains(transaction.txId)) {
            confirmedTxIdTransactionReply[transaction.txId] = transactionInfo
            if (transactionInfo.confirmations > 0) {
                pendingTxIdTransaction.remove(transaction.txId)
                transactionsForMonitoring.remove(transaction.txId)
            }
        }
    }

    fun addTransaction(transaction: Transaction): Boolean {
        val transactions = pendingAddressTransaction[transaction.inputAddress.toString()]
        if (!transactions.isNullOrEmpty()) {
            transactions.add(transaction)
            pendingTxIdTransaction[transaction.txId] = transaction
        } else {
            val newList = ArrayList<Transaction>()
            newList.add(transaction)
            pendingTxIdTransaction[transaction.txId] = transaction
            pendingAddressTransaction[transaction.inputAddress.toString()] = newList
        }
        return true
    }

    fun getTransaction(txId: Sha256Hash): Transaction? {
        return pendingTxIdTransaction[txId]
    }

    fun getPendingSignatureIndexForAddress(address: String): Long? {
        val transactions = pendingAddressTransaction[address]
        return if (!transactions.isNullOrEmpty()) {
            transactions[transactions.size - 1].getSignatureIndex()
        } else {
            null
        }
    }
}
